typing_indicator: Update APIs for stream typing notifications.

prakhar1144 opened the pull request:
[CZO discussion](https://chat.zulip.org/#narrow/stream/378-api-design/topic/stream.20typing.20indicator/near/1649887)

> For https://zulip.com/api/get-events#typing-start, I think we should change the value "private" to "direct" --

> We could also remove the "private" compatibility support in [/api/set-typing-status](https://chat.zulip.org/api/set-typing-status) -- assuming we confirm that the mobile and terminal clients don't actually send the value "private"


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

---

<details><summary>set-typing-status</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/c22b129d-e535-4cc3-a7f4-cc1a80b52f8c">
</img>
</details> 

<details><summary>get-events#typing-start</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/f70adf97-1a73-4c59-adc6-5cc66f866903">
</img>
</details> 

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented:
TODO -- [decision pending
](https://chat.zulip.org/#narrow/stream/378-api-design/topic/stream.20typing.20indicator/near/1650167)

> Oh, here's one other change that I think we can make for free and would be a good cleanup:

> In [/api/set-typing-status](https://chat.zulip.org/api/set-typing-status), when type is stream, we still expect to to be a list:
  
> For "stream" type it is a single element list containing ID of stream in which the message is being typed. That seems a bit silly — should just be an int, the stream ID.
  
> Perhaps better yet: the stream ID could be a parameter with a different name, like stream_id instead of to. That way it can have type int while to remains of type list-of-int.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -16538,7 +16538,6 @@ paths:
             enum:
               - direct
               - stream
-              - private
```
We need to add a note above `**Changes**: In Zulip 7.0 (feature level 174)` that `private` stopped being supported at this feature level. (I think you can use the same feature level for this commit as the previous one). And this should be mentioned in the API changelog too.

prakhar1144 replied:
<details><summary>set-typing-status</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/d8fac7ca-17f4-461c-a46d-7e30ee819e28">
</img>
</details>

timabbott replied:
Looks great!

timabbott commented:
These changes lgtm aside from that detail.

prakhar1144 commented:
Addressed the review comments.

We should wait to merge this if we plan to [make this change](https://github.com/zulip/zulip/pull/26958#issuecomment-1742175858) after discussion.

timabbott commented:
@prakhar1144 I think we decided what we wanted to do with that last detail, right?

prakhar1144 commented:
I was awaiting if @gnprice has any feedback, as [you @-mentioned him](https://chat.zulip.org/#narrow/stream/378-api-design/topic/stream.20typing.20indicator/near/1652618)

I will implement this because the only pending decision would be: the name being `to` or `stream_id` (greg earlier suggested)

We all agree on the value being `int` or `List[int]`

prakhar1144 commented:
Updates in the recent push:

* Worked on the [todo comment](https://github.com/zulip/zulip/pull/26958#issuecomment-1742175858)
* Last 3 commits are new

prakhar1144 commented on zerver/lib/request.py:
```diff
@@ -491,3 +491,28 @@ def _wrapped_req_func(
+        if not isinstance(user_id, int):
+            raise JsonableError(_("Recipient list may only contain user IDs"))
+
+    return list(set(user_ids))
```
Should we put these functions in a new file (These functions are used in `set typing status` & `scheduled messages` codepath) ? 

Among existing files this looked the best option to me as they are mainly used in `views` to extract request data.

timabbott replied:
Yeah, this file is meant to be just for the `has_request_variables` system, which we're planning to delete.

I think putting these in a new `zerver/lib/recipient_parsing.py`, and then as a follow-up moving `extract_private_recipients` / `extract_stream_indicator` to that file, would be a decent strategy. And we can make a new `zerver/tests/test_recipient_parsing` for that new file as well.

prakhar1144 replied:
Made the changes -- 4th commit.

prakhar1144 commented:
Updates in the recent push:

* Worked on the [todo comment](https://github.com/zulip/zulip/pull/26958#issuecomment-1742175858)
* Last 3 commits are new

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,13 @@ format used by the Zulip server that they are interacting with.
+* [`POST /typing`](/api/set-typing-status): When the `type` of the message
+  being composed is `"stream"`, changed the `to` parameter to accept the
+  ID of the stream in which the message is being typed. Previously, it
+  accepted a single-element list containing the ID of the stream.
```
This seems like it should be feature level 215 as well. And maybe it's worth a note just after the bullets for that feature level that "Note that stream typing notifications were not enabled in any Zulip client prior to feature level 215.".

prakhar1144 replied:
Made the changes.

Additionally added a line in `zulip.yaml` as [discussed](https://chat.zulip.org/#narrow/stream/378-api-design/topic/stream.20typing.20indicator/near/1658910):

"Clients shouldn't care about the APIs prior to Zulip 8.0 (feature level 215) for stream typing notifications, as no client actually implemented the previous API for those."

timabbott commented:
Posted a few comments, but overall this looks like it should merge after the next round of revisions.

prakhar1144 commented:
@timabbott 

This is ready to review.

Updates in the recent push:
* Addressed the review comments.

We should merge this & #26904 in order; to make sure that the [comments in the api docs](https://github.com/zulip/zulip/pull/26958#discussion_r1350763830) stays relevant.

timabbott commented:
Merged, thanks @prakhar1144!
